php/Blockonomics.php

```diff
@@ -238,4 +292,267 @@ public function testSetup()
+                    exit();
+                }
+                $price = $responseObj->price;
+                $price = $price * 100/(100+get_option('blockonomics_margin', 0));
+            }else{
+                $price = 1;
+            }
+            $satoshi = intval(round(1.0e8*$wc_order->get_total()/$price));
+            $status = -1;
+        }
+        $order = array(
+                'value'              => $wc_order->get_total(),
+                'currency'           => get_woocommerce_currency(),
+                'order_id'           => $order_id,
+                'status'             => $status,
+                'satoshi'            => $satoshi,
+                'crypto'             => $crypto,
+                'timestamp'          => $timestamp,
+                'addr'               => $address,
+                'txid'               => ''
+        );
+
+        $orders[$order_id][$address] = $order;
+        update_option('blockonomics_orders', $orders);
+        update_post_meta($order_id, $crypto .'_address', $address);
+        return $order;
+    }
+
+    // Get the order info by id and crypto
+    public function get_order_info($order_id, $crypto){
+        $order = $this->process_order($order_id, $crypto);
+        header("Content-Type: application/json");
+        exit(json_encode($order));
+    }
+
+    // Process the blockonomics callback
+    public function process_callback($secret, $addr, $status, $value, $txid){
+        $callback_secret = get_option("blockonomics_callback_secret");
+        if ($callback_secret  && $callback_secret == $secret) {
+            $orders = get_option('blockonomics_orders');
+            // fetch the order id by bitcoin address
+            foreach($orders as $id => $order){
+                if(isset($order[$addr])){
+                    $order_id = $id;
+                    break;
+                }
+            }
+            if ($order_id){
+                $order = $orders[$order_id][$addr];
+                $wc_order = new WC_Order($order_id);
+                $existing_status = $order['status'];
+                $timestamp = $order['timestamp'];
+                $time_period = get_option("blockonomics_timeperiod", 10) *60;
+                $network_confirmations=get_option("blockonomics_network_confirmation",2);
+                if ($status == 0 && time() > $timestamp + $time_period) {
+                    $minutes = (time() - $timestamp)/60;
+                    $wc_order->add_order_note(__("Warning: Payment arrived after $minutes minutes. Received". $order['crypto'] ."may not match current ". $order['crypto'] ." price", 'blockonomics-bitcoin-payments'));
+                }
+                elseif ($status >= $network_confirmations && !metadata_exists('post',$wc_order->get_id(),'paid_'. $order['crypto'] .'_amount') )  {
+                    update_post_meta($wc_order->get_id(), 'paid_'. $order['crypto'] .'_amount', $value/1.0e8);
```

Comment:
This is a bit confusing on why were are checking metadata etc. Create a method

*recordPayment()* : Params value, order_id and cryto. If payment is already recorded returns False else records payment and return True.

Reply:
Refactored in 62c1ec2